### PR TITLE
Add support for Express Route Provider APIs

### PR DESCRIPTION
**Pull request by gimotwanMSFT**
A new top level NRP resource ExpressRouteCrossConnection is being
    introduced with this change. This resource shadows the customer
    ExpressRouteCircuit resouce in the customer subscription and enables
    the     provider to perform CRUD operations on some of the
    sub-resources, such as peerings on the ExpressRoute Circuit

    * update readme.md to include the complete path for the 2018-02-01
      folder
    * add examples
    * add ExpressRouteCrossConnection as a top level resource

### CODE REVIEW

File: specification/network/resource-manager/Microsoft.Network/stable/2018-02-01/expressRouteCrossConnection.json [hunk 1]
```diff
@@ -0,0 +1,738 @@
...
+      }
+    }
+  },
+  "paths": {
+    "/subscriptions/{subscriptionId}/providers/Microsoft.Network/expressRouteCrossConnections": {
+      "get": {
+        "tags": [
+          "ExpressRouteCrossConnections"
+        ],
+        "operationId": "ExpressRouteCrossConnections_ListAll",
```
Comment (dsgouda) [inline specification/network/resource-manager/Microsoft.Network/stable/2018-02-01/expressRouteCrossConnection.json]: Recommended operationId for list operations is `ExpressRouteCrossConnections_List`

File: specification/network/resource-manager/Microsoft.Network/stable/2018-02-01/expressRouteCrossConnection.json [hunk 2]
```diff
@@ -0,0 +1,738 @@
...
+          "nextLinkName": "nextLink"
+        }
+      }
+    },
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Network/expressRouteCrossConnections": {
+      "get": {
+        "tags": [
+          "ExpressRouteCrossConnections"
+        ],
+        "operationId": "ExpressRouteCrossConnectionsByResourceGroup_List",
```
Comment (dsgouda) [inline specification/network/resource-manager/Microsoft.Network/stable/2018-02-01/expressRouteCrossConnection.json]: recommended operationId is `ExpressRouteCrossConnections_ListByResourceGroup`

File: specification/network/resource-manager/Microsoft.Network/stable/2018-02-01/expressRouteCrossConnection.json [hunk 3]
```diff
@@ -0,0 +1,738 @@
...
+          "required": true,
+          "type": "string",
+          "description": "The name of the cross connection."
+        },
+        {
+          "name": "parameters",
+          "in": "body",
+          "required": true,
+          "schema": {
+          "$ref": "./network.json#/definitions/TagsObject"
```
Comment (dsgouda) [inline specification/network/resource-manager/Microsoft.Network/stable/2018-02-01/expressRouteCrossConnection.json]: nit: fix indentation
Shouldn't this parameter be named `tags` instead?

File: specification/network/resource-manager/Microsoft.Network/stable/2018-02-01/expressRouteCrossConnection.json [hunk 4]
```diff
@@ -0,0 +1,738 @@
...
+        "responses": {
+        "200": {
+          "description": "Update successful. The operation returns the resulting ExpressRouteCrossConnection resource.",
+          "schema": {
+          "$ref": "#/definitions/ExpressRouteCrossConnection"
+          }
+        }
+        },
+        "x-ms-examples": {
+        "Update Express Route CrossConnection Tags": {
```
Comment (dsgouda) [inline specification/network/resource-manager/Microsoft.Network/stable/2018-02-01/expressRouteCrossConnection.json]: Remove spaces (basic Json keys' semantics)

File: specification/network/resource-manager/Microsoft.Network/stable/2018-02-01/expressRouteCrossConnection.json [hunk 5]
```diff
@@ -0,0 +1,738 @@
...
+        },
+        "x-ms-long-running-operation": true
+      }
+    },
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Network/expressRouteCrossConnections/{crossConnectionName}/peerings/{peeringName}/arpTables/{devicePath}": {
+      "post": {
+        "tags": [
+          "ExpressRouteCrossConnectionArpTable"
+        ],
+        "operationId": "ExpressRouteCrossConnections_ListArpTable",
```
Comment (dsgouda) [inline specification/network/resource-manager/Microsoft.Network/stable/2018-02-01/expressRouteCrossConnection.json]: Rename as `ExpressRouteCircuitsArpTable_list`

File: specification/network/resource-manager/Microsoft.Network/stable/2018-02-01/expressRouteCrossConnection.json [hunk 6]
```diff
@@ -0,0 +1,738 @@
...
+        },
+        "x-ms-long-running-operation": true
+      }
+    },
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Network/expressRouteCrossConnections/{crossConnectionName}/peerings/{peeringName}/routeTables/{devicePath}": {
+      "post": {
+        "tags": [
+          "ExpressRouteCrossConnectionRouteTable"
+        ],
+        "operationId": "ExpressRouteCrossConnections_ListRoutesTable",
```
Comment (dsgouda) [inline specification/network/resource-manager/Microsoft.Network/stable/2018-02-01/expressRouteCrossConnection.json]: rename as `ExpressRouteCircuitsRoutesTable_List`

File: specification/network/resource-manager/Microsoft.Network/stable/2018-02-01/expressRouteCrossConnection.json [hunk 7]
```diff
@@ -0,0 +1,738 @@
...
+              "description": "Request successful. The operation returns the resulting ExpressRouteCrossConnectionsRouteTable resource.",
+              "schema": {
+                "$ref": "./expressRouteCircuit.json#/definitions/ExpressRouteCircuitsRoutesTableListResult"
+              }
+          },
+          "202": {
+            "description": "Accepted and the operation will complete asynchronously."
+          }
+        },
+        "x-ms-long-running-operation": true
```
Comment (dsgouda) [inline specification/network/resource-manager/Microsoft.Network/stable/2018-02-01/expressRouteCrossConnection.json]: Is there a reason why this list operation is long running

File: specification/network/resource-manager/Microsoft.Network/stable/2018-02-01/expressRouteCrossConnection.json [hunk 8]
```diff
@@ -0,0 +1,738 @@
...
+    "ExpressRouteCircuitReference": {
+      "properties": {
+        "id": {
+          "type":"string",
+          "description": "Corresponding Express Route Circuit Id."
+        }
+      }
+    },
+    "ExpressRouteCrossConnectionServiceProviderProperties": {
+      "properties": {
```
Comment (dsgouda) [inline specification/network/resource-manager/Microsoft.Network/stable/2018-02-01/expressRouteCrossConnection.json]: Are any of these `readonly`?

File: specification/network/resource-manager/Microsoft.Network/stable/2018-02-01/expressRouteCrossConnection.json [hunk 9]
```diff
@@ -0,0 +1,738 @@
...
+        "serviceProviderProperties": {
+          "$ref": "#/definitions/ExpressRouteCrossConnectionServiceProviderProperties",
+          "description": "The ServiceProviderProperties."
+        },
+        "expressRouteCircuit": {
+          "$ref": "#/definitions/ExpressRouteCircuitReference",
+          "description": "The ExpressRouteCircuit"
+        },
+        "serviceProviderProvisioningState": {
+          "type": "string",
```
Comment (dsgouda) [inline specification/network/resource-manager/Microsoft.Network/stable/2018-02-01/expressRouteCrossConnection.json]: Provisioning states are usually readonly. Please confirm.

File: specification/network/resource-manager/Microsoft.Network/stable/2018-02-01/expressRouteCrossConnection.json [hunk 10]
```diff
@@ -0,0 +1,738 @@
...
+            "modelAsString": true
+          }
+        },
+        "serviceProviderNotes": {
+          "type": "string",
+          "description": "Additional read only notes set by the connectivity provider."
+        },
+        "provisioningState": {
+          "type": "string",
+          "description": "Gets the provisioning state of the public IP resource. Possible values are: 'Updating', 'Deleting', and 'Failed'."
```
Comment (dsgouda) [inline specification/network/resource-manager/Microsoft.Network/stable/2018-02-01/expressRouteCrossConnection.json]: Definitely readonly

File: specification/network/resource-manager/Microsoft.Network/stable/2018-02-01/expressRouteCrossConnection.json [hunk 11]
```diff
@@ -0,0 +1,738 @@
...
+          "type": "string",
+          "description": "Gets a unique read-only string that changes whenever the resource is updated."
+        }
+      },
+      "allOf": [
+        {
+          "$ref": "./network.json#/definitions/Resource"
+        }
+      ],
+        "description": "ExpressRouteCrossConnection resource"
```
Comment (dsgouda) [inline specification/network/resource-manager/Microsoft.Network/stable/2018-02-01/expressRouteCrossConnection.json]: nit: indentation

File: specification/network/resource-manager/Microsoft.Network/stable/2018-02-01/expressRouteCrossConnection.json [hunk 12]
```diff
@@ -0,0 +1,738 @@
...
+        "value": {
+          "type": "array",
+          "items": {
+            "$ref": "#/definitions/ExpressRouteCrossConnection"
+          },
+          "description": "A list of ExpressRouteCrossConnection resources."
+        },
+        "nextLink": {
+          "type": "string",
+          "description": "The URL to get the next set of results."
```
Comment (dsgouda) [inline specification/network/resource-manager/Microsoft.Network/stable/2018-02-01/expressRouteCrossConnection.json]: mark as readonly

File: specification/network/resource-manager/Microsoft.Network/stable/2018-02-01/expressRouteCrossConnection.json [hunk 13]
```diff
@@ -0,0 +1,738 @@
...
+      "description": "Response for ListExpressRouteCrossConnection API service call."
+    }
+  },
+  "parameters": {
+    "SubscriptionIdParameter": {
+      "name": "subscriptionId",
+      "in": "path",
+      "required": true,
+      "type": "string",
+    "description": "The subscription credentials which uniquely identify the Microsoft Azure subscription. The subscription ID forms part of the URI for every service call."
```
Comment (dsgouda) [inline specification/network/resource-manager/Microsoft.Network/stable/2018-02-01/expressRouteCrossConnection.json]: nit: indentation

**Review by dsgouda — CHANGES_REQUESTED** [review]
Posted a few comments. Please check CI jobs and resolve issues if any

File: specification/network/resource-manager/Microsoft.Network/stable/2018-02-01/expressRouteCrossConnection.json (hunk 3 above)
Comment (gimotwanMSFT) [reply specification/network/resource-manager/Microsoft.Network/stable/2018-02-01/expressRouteCrossConnection.json]: Fixed indentation. The naming is consistent with corresponding operation on ExpressRoute Circuit

File: specification/network/resource-manager/Microsoft.Network/stable/2018-02-01/expressRouteCrossConnection.json (hunk 7 above)
Comment (gimotwanMSFT) [reply specification/network/resource-manager/Microsoft.Network/stable/2018-02-01/expressRouteCrossConnection.json]: The Cross Connection resource shadows the corresponding Express Route Circuit resource. The list route table operation would fetch the contents for the route tables from the device for multiple tables. Further each table may have several records.

File: specification/network/resource-manager/Microsoft.Network/stable/2018-02-01/expressRouteCrossConnection.json (hunk 9 above)
Comment (gimotwanMSFT) [reply specification/network/resource-manager/Microsoft.Network/stable/2018-02-01/expressRouteCrossConnection.json]: The serviceProviderProvisioning state is writeable by the ExpressRoute Provider subscription.

File: specification/network/resource-manager/Microsoft.Network/stable/2018-02-01/expressRouteCrossConnection.json (hunk 8 above)
Comment (gimotwanMSFT) [reply specification/network/resource-manager/Microsoft.Network/stable/2018-02-01/expressRouteCrossConnection.json]: Some are. Updated in the subsequent commit.

File: specification/network/resource-manager/Microsoft.Network/stable/2018-02-01/expressRouteCrossConnection.json (hunk 12 above)
Comment (gimotwanMSFT) [reply specification/network/resource-manager/Microsoft.Network/stable/2018-02-01/expressRouteCrossConnection.json]: Should the nextLink record be marked readOnly ? A quick browse through some of the other files does not show nextLink marked as readOnly. Please advise and I will update.

File: specification/network/resource-manager/Microsoft.Network/stable/2018-02-01/expressRouteCrossConnection.json (hunk 3 above)
Comment (gimotwanMSFT) [reply specification/network/resource-manager/Microsoft.Network/stable/2018-02-01/expressRouteCrossConnection.json]: Thank you for helping with the reviews.

File: specification/network/resource-manager/Microsoft.Network/stable/2018-02-01/expressRouteCrossConnection.json (hunk 8 above)
Comment (dsgouda) [reply specification/network/resource-manager/Microsoft.Network/stable/2018-02-01/expressRouteCrossConnection.json]: Similarly, please go through all the models and their properties in the spec and verify their `readonly`-ness

File: specification/network/resource-manager/Microsoft.Network/stable/2018-02-01/expressRouteCrossConnection.json (hunk 12 above)
Comment (dsgouda) [reply specification/network/resource-manager/Microsoft.Network/stable/2018-02-01/expressRouteCrossConnection.json]: Yes, please mark this as readonly

File: specification/network/resource-manager/Microsoft.Network/stable/2018-02-01/expressRouteCrossConnection.json (hunk 3 above)
Comment (dsgouda) [reply specification/network/resource-manager/Microsoft.Network/stable/2018-02-01/expressRouteCrossConnection.json]: `parameters` is still to generic, please name it something more meaningful

File: specification/network/resource-manager/Microsoft.Network/stable/2018-02-01/expressRouteCrossConnection.json [hunk 14]
```diff
@@ -0,0 +1,739 @@
...
+            "description": "Update successful. The operation returns the resulting ExpressRouteCrossConnection resource."
+          }
+        },
+        "x-ms-long-running-operation": true
+      },
+      "patch": {
+        "tags": [
+        "ExpressRouteCrossConnections"
+        ],
+        "operationId": "ExpressRouteCrossConnections_UpdateTags",
```
Comment (dsgouda) [inline specification/network/resource-manager/Microsoft.Network/stable/2018-02-01/expressRouteCrossConnection.json]: Rename this as `ExpressRouteCrossConnectionsTags_Update`

**Review by dsgouda — CHANGES_REQUESTED** [review]
A few more comments.

File: specification/network/resource-manager/Microsoft.Network/stable/2018-02-01/expressRouteCrossConnection.json [hunk 15]
```diff
@@ -0,0 +1,744 @@
...
+          "description": "The list of peerings."
+        }
+      },
+      "description": "Properties of ExpressRouteCrossConnection."
+    },
+    "ExpressRouteCrossConnection": {
+      "properties": {
+        "properties": {
+          "x-ms-client-flatten": true,
+          "$ref": "#/definitions/ExpressRouteCrossConnectionPropertiesFormat"
```
Comment (dsgouda) [inline specification/network/resource-manager/Microsoft.Network/stable/2018-02-01/expressRouteCrossConnection.json]: nit: please rename this as `ExpressRouteCrossConnectionProperties`

**Review by dsgouda — CHANGES_REQUESTED** [review]
A minor comment and please delete the .swp file

File: specification/network/resource-manager/Microsoft.Network/stable/2018-02-01/expressRouteCrossConnection.json [hunk 16]
```diff
@@ -0,0 +1,743 @@
...
+        "responses": {
+        "200": {
+          "description": "Update successful. The operation returns the resulting ExpressRouteCrossConnection resource.",
+          "schema": {
+          "$ref": "#/definitions/ExpressRouteCrossConnection"
+          }
+        }
+        },
+        "x-ms-examples": {
+        "UpdateExpressRouteCrossConnectionTags": {
```
Comment (dsgouda) [inline specification/network/resource-manager/Microsoft.Network/stable/2018-02-01/expressRouteCrossConnection.json]: nit: fix indentation

File: specification/network/resource-manager/Microsoft.Network/stable/2018-02-01/expressRouteCrossConnection.json [hunk 17]
```diff
@@ -0,0 +1,743 @@
...
+          },
+          {
+            "$ref": "#/parameters/SubscriptionIdParameter"
+          }
+        ],
+        "responses": {
+          "200": {
+            "description": "Request successful. The operation returns the resulting ExpressRouteCrossConnectionsRouteTableSummary resource.",
+            "schema": {
+            "$ref": "./expressRouteCircuit.json#/definitions/ExpressRouteCircuitsRoutesTableSummaryListResult"
```
Comment (dsgouda) [inline specification/network/resource-manager/Microsoft.Network/stable/2018-02-01/expressRouteCrossConnection.json]: nit: Fix indentation

File: specification/network/resource-manager/Microsoft.Network/stable/2018-02-01/expressRouteCrossConnection.json [hunk 18]
```diff
@@ -0,0 +1,743 @@
...
+          {
+            "$ref": "#/parameters/ApiVersionParameter"
+          },
+          {
+            "$ref": "#/parameters/SubscriptionIdParameter"
+          }
+        ],
+        "responses": {
+          "200": {
+              "description": "Request successful. The operation returns the resulting ExpressRouteCrossConnectionsRouteTable resource.",
```
Comment (dsgouda) [inline specification/network/resource-manager/Microsoft.Network/stable/2018-02-01/expressRouteCrossConnection.json]: Indentation

**Review by dsgouda — CHANGES_REQUESTED** [review]

**Review by dsgouda — CHANGES_REQUESTED** [review]
There 2 more violations you need to fix.
- Not all operations have corresponding examples. Please provide them.
- Please verify whether information related to the service appears in the List_Operations API for network

**Comment by dsgouda** [comment]
A few examples are still missing, please check [here](https://travis-ci.org/Azure/azure-rest-api-specs/jobs/347035430)

**Comment by gimotwanMSFT** [comment]
How do I verify the following ?
Please verify whether information related to the service appears in the List_Operations API for network

**Comment by dsgouda** [comment]
@gimotwanMSFT disregard that comment

**Review by dsgouda — APPROVED** [review]
LGTM

**Pull request merged**